libcaf_core/caf/async/future.hpp

```diff
@@ -0,0 +1,138 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include "caf/async/execution_context.hpp"
+#include "caf/async/fwd.hpp"
+#include "caf/detail/async_cell.hpp"
+#include "caf/disposable.hpp"
+#include "caf/error.hpp"
+#include "caf/sec.hpp"
+
+namespace caf::async {
+
+/// Provides an interface for accessing the result of an asynchronous
+/// computation on an asynchronous @ref execution_context.
+template <class T>
+class bound_future {
+public:
+  friend class future<T>;
+
+  bound_future() noexcept = default;
+  bound_future(bound_future&&) noexcept = default;
+  bound_future(const bound_future&) noexcept = default;
+  bound_future& operator=(bound_future&&) noexcept = default;
+  bound_future& operator=(const bound_future&) noexcept = default;
+
+  /// Retrieves the result at some point in the future and then calls either
+  /// @p on_success  if the asynchronous operation generated a result or
+  /// @p on_error if the asynchronous operation resulted in an error.
+  template <class OnSuccess, class OnError>
+  disposable then(OnSuccess on_success, OnError on_error) {
+    static_assert(std::is_invocable_v<OnSuccess, const T&>);
+    static_assert(std::is_invocable_v<OnError, const error&>);
+    auto cb = [cp = cell_, f = std::move(on_success), g = std::move(on_error)] {
+      // Note: no need to lock the mutex. Once the cell has a value and actions
+      // are allowed to run, the value is immutable.
+      switch (cp->value.index()) {
+        default:
+          g(make_error(sec::broken_promise, "future found an invalid value"));
+          break;
+        case 1:
+          f(static_cast<const T&>(std::get<T>(cp->value)));
+          break;
+        case 2:
+          g(static_cast<const error&>(std::get<error>(cp->value)));
+      }
+    };
+    auto cb_action = make_single_shot_action(std::move(cb));
```

Comment:
The result of this change is that even if multiple promises of the same future try to set a value, only (the most recent?) one is propagated?